File: 494_Target Sum.java
// 时间复杂度：O(nm)，其中 n 为 nums 的长度，m 为 nums 的元素和减去 ∣target∣。
// 空间复杂度：O(m)

class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        // Method 1 (slow): binary tree - each branch is eith + or -
        // Method 2: 0-1背包 (DP)
        // sum of all selected positive nums[i] = p
        // the absolute value for sum of all selected negative nums[i] = q
        // p + q = complete array sum s
        // p - q = target t
        // Therefore, p = (t + s) / 2 => bag size = p => how many ways to fill the backpack

        int sum = 0;
        for (int n : nums) sum += n;
        // cut branch
        if ((sum + target) % 2 != 0 || (sum + target) < 0) return 0;

        // 0-1背包问题
        int bagSize = (sum + target) / 2;
        int[] dp = new int[bagSize + 1];
        dp[0] = 1;

        for (int i : nums) {
            for (int j = bagSize; j >= i; j--) {
                if (dp[j - i] > 0) dp[j] += dp[j-i];
            }
        }

        return dp[bagSize];
    }
}

File: 494_Target Sum.py
''' Assume the set p for plus and the set m for minus, 
sum(p) - sum(m) = target
sum(p) + sum(m) = total
->
sum(p) = (target + total) >> 1 <- fill this bag '''

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        size = sum(nums) + target
        if size % 2 != 0 or size < 0: return 0

        size >>= 1
        dp = [1] + [0] * size

        for n in nums:
            for i in range(size, n-1, -1):
                dp[i] += dp[i - n]
        return dp[-1]